File: events/2015/02/solution.ts
import { AOCDayResults } from "@repo/utils-ts/returnType";
import { readInputAsArray } from "@repo/utils-ts/readInput";
import { measurePerformance } from "@repo/utils-ts/measurePerformance";

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
function part1(input: string[]): number | undefined {
  let totalArea = 0;

  // Work through every present
  for (const present of input) {
    // Get the dimensions of the present
    const [length, width, height] = present.split('x').map(Number);

    // Calculate the area for the present
    let area = (2 * length * width) + (2 * width * height) + (2 * height * length);

    // Find the smallest side
    let smallestSide = Math.min((length * width), (width * height), (height * length));
    area += smallestSide;

    // Add the area to the total area
    totalArea += area;
  }

  // Return the total area
  return totalArea;
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
function part2(input: string[]): number | undefined {
  let totalRibbon = 0;

  // Work through every present
  for (const present of input) {
    // Get the dimensions of the present
    const [length, width, height] = present.split('x').map(Number);

    const perimeter = 2 * Math.min(length + width, width + height, height + length);

    // Calculate the volume
    const volume = length * width * height;

    // Add the length to the total ribbon
    totalRibbon += perimeter + volume;
  }

  // Return the total area
  return totalRibbon;
}


// ----------------------------------------------------------------------------------------------------
// | Solve the puzzle
// ----------------------------------------------------------------------------------------------------
function solve() {
  // Get the arguments
  const args = process.argv.slice(2);
  const fileName = args[0] || "input.txt";

  // Start the timer
  const startTimeStamp = new Date(Date.now()).toISOString();
  const startTime = performance.now();

  // Read the input as an array of strings
  const input = readInputAsArray(fileName);

 // Run the parts
 const p1Result = measurePerformance(() => part1(input));
 const p2Result = measurePerformance(() => part2(input));

 // End the timer
 const endTimeStamp = new Date(Date.now()).toISOString();
 const endTime = performance.now();
 const duration = endTime - startTime;

 // Return the results
 return {
  Year: 2015,
  Day: 2,
  Part1: p1Result,
  Part2: p2Result,
  Duration: Math.round(duration * 1000000), // Convert to nanoseconds
  Timestamp: {
    Start: new Date(startTimeStamp).toISOString(),
    End: new Date(endTimeStamp).toISOString(),
  },
 } as AOCDayResults;
  
}
const results = solve();
console.log(JSON.stringify(results, null, 2));

File: events/2015/02/solution.py
import sys
import time
from datetime import timedelta, datetime
from pathlib import Path
import json

# Add workspace root to path for imports
workspace_root = Path(__file__).parent.parent.parent.parent
sys.path.insert(0, str(workspace_root))

from utils.python.types import AOCDayResults, AOCTimestamp, AOCPartResult
from utils.python.input import readInputAsArray
from utils.python.performance import measurePerformance

# ----------------------------------------------------------------------------------------------------
# | Part 1
# ----------------------------------------------------------------------------------------------------
def part1(input: list[str]) -> int:
  total_area = 0

  for present in input:
    length, width, height = [int(x) for x in present.split("x")]

    area = (2 * length * width) + (2 * width * height) + (2 * height * length)

    smallestSide = min((length * width), (width * height), (height * length))
    area += smallestSide

    total_area += area

  return total_area

# ----------------------------------------------------------------------------------------------------
# | Part 2
# ----------------------------------------------------------------------------------------------------
def part2(input: list[str]) -> int:
  total_ribbon = 0

  for present in input:
    length, width, height = [int(x) for x in present.split("x")]

    perimeter = 2 * min(length + width, width + height, height + length)
    volume = length * width * height

    total_ribbon += perimeter + volume

  return total_ribbon

# ----------------------------------------------------------------------------------------------------
# | Main Function
# ----------------------------------------------------------------------------------------------------
def main():
  # Get the arguments
  args = sys.argv[1:]
  fileName = args[0] if args else "input.txt"

  # Start the timer
  start_timestamp = datetime.now().isoformat()
  start_time = time.monotonic()

  # Read the input to an array of strings
  input = readInputAsArray(fileName)

  ## Run the parts
  p1Result = measurePerformance(part1, input)
  p2Result = measurePerformance(part2, input)

  # End the timer
  end_time = time.monotonic()
  end_timestamp = datetime.now().isoformat()
  duration = timedelta(seconds=end_time - start_time).total_seconds()*1000000000 # Convert to nanoseconds

  # Return the results
  return AOCDayResults(
    Year=2015,
    Day=2,
    Part1=p1Result,
    Part2=p2Result,
    Duration=duration,
    Timestamp=AOCTimestamp(
      Start=start_timestamp,
      End=end_timestamp,
    ),
  )



if __name__ == "__main__":
  results = main()
  print(json.dumps(results))